src/key_store/mod.rs

```diff
@@ -0,0 +1,222 @@
+//! to consume a `KeyPackageBundle` instance. Note, that in contrast to
+//! `CredentialBundle` instances, `KeyPackageBundle` instances should not be
+//! used across groups.
+//!
+//! This design is temporary and only until the `ManagedGroup` is refactored to
+//! access its `KeyPackageBundle` via the `KeyStore`. Once this is the case, the
+//! `take_key_package_bundle` will be deprecated in favor of a
+//! `get_key_package_bundle`, which only returns a reference to the
+//! `KeyPackageBundle`.
+//!
+//! # Example
+//!
+//! A simple example for the generation and the retrieval of a
+//! `CredentialBundle` and a `KeyPackageBundle`.
+//!
+//! ```
+//! use openmls::prelude::*;
+//!
+//! let key_store = KeyStore::default();
+//!
+//! let ciphersuite_name = CiphersuiteName::MLS10_128_DHKEMX25519_AES128GCM_SHA256_Ed25519;
+//!
+//! // Generate a credential bundle with the matching signature scheme.
+//! let alice_credential = key_store
+//!     .generate_credential(
+//!         "Alice".into(),
+//!         CredentialType::Basic,
+//!         SignatureScheme::from(ciphersuite_name),
+//!     )
+//!     .unwrap();
+//!
+//! // Generate a key package bundle.
+//! let alice_key_package = key_store
+//!     .generate_key_package(&[ciphersuite_name], &alice_credential, vec![])
+//!     .unwrap();
+//!
+//! // Create a group with the previously generated credential and key package.
+//! let managed_group_config = ManagedGroupConfig::new(
+//!     HandshakeMessageFormat::Plaintext,
+//!     UpdatePolicy::default(),
+//!     0,
+//!     0,
+//!     ManagedGroupCallbacks::default(),
+//! );
+//!
+//! let alice_group = ManagedGroup::new(
+//!     &key_store,
+//!     &managed_group_config,
+//!     GroupId::from_slice(b"Test Group"),
+//!     &alice_key_package.hash(),
+//! )
+//! .unwrap();
+//! ```
+//!
+//! # Future Work
+//!
+//! A more detailed roadmap can be found in issue #337, but generally the plan is to
+//! * move ownership of all `KeyPackageBundle` instances into the `KeyStore` and
+//!   change the API from `take_` to `get_`
+//! * keep track of which key material is (still) in use and where, as well as if it expires and when
+//! * extend the API to allow for deletion of unused key material, expired or otherwise
+//! * add a persistence layer
+
+use std::{
+    collections::HashMap,
+    sync::{RwLock, RwLockReadGuard},
+};
+
+use crate::{
+    ciphersuite::{CiphersuiteName, SignaturePublicKey, SignatureScheme},
+    credentials::{Credential, CredentialBundle, CredentialType},
+    extensions::Extension,
+    key_packages::{KeyPackage, KeyPackageBundle},
+};
+
+pub mod errors;
+
+#[cfg(test)]
+mod test_key_store;
+
+pub use errors::KeyStoreError;
+
+#[derive(Debug, Default)]
+/// The `KeyStore` contains `CredentialBundle`s and `KeyPackageBundle`s and
+/// makes them available via information in the corresponding `Credential` and
+/// `KeyPackage` instances.
+pub struct KeyStore {
```

Comment:
Looks like we currently don't have a way to actually delete anything here? I think we can go without in the first version but should document it. Because this will also be necessary for testing of all the async things it would be nice to have it earlier. But I'm fine without in the first version.

Reply:
Yes, I've not added that yet. I thought I'd add it once we have a way to keep track of `Credential` usage, to make sure that we can safely delete `CredentialBundle` instances. Not sure when that is going to happen, though.